### ISSUE TITLE
Potential bug in get_disparity_smoothness

### ISSUE THREAD

**Issue by yzcjtr** [mrharicot/monodepth#46]
Hi,

I think I've found a bug in the function get_disparity_smoothness in monodepth_model.py.  The last line concats two lists instead of adding them elementwise. The actual return of the function is a list of 8 tensors of shape B*H*W*1. Thus the use of the function is inappropriate.

**Comment by mrharicot** [comment]
Looks like you're right!
Thanks for spotting this.
I'll think of a way to fix this.

**Comment by ninception** [comment]
Hey! even I noticed this bug. But when I try to add them up in a single line of list comprehension, I get an error stating the sizes for gradients are different. This is somewhat expected since the gradient_x is of size (H, W-1) while the gradient_y is of size (H-1, W) as a consequence of your function to get the X and Y image gradients. 

Even on fixing that (using some crude hack basically interpolating for excluded border row and column), while the adding the weighted X and Y gradients in the list comprehension, I was getting another error stating the batch sizes are different (8 and 7 resp.) for gradients_x[i] and gradients_y[i]. 

For now,  I am just crudely adding them whenever batch sizes matches and where they don't match I just give the weighted value for the X gradients.

**Comment by a-jahani** [comment]
Does this issue still exist? Any solution code suggested? Thanks in advance

**Comment by shubhamwagh** [comment]
Yeah, this issue is not yet solved. @mrharicot  if you could give some insight on this.

**Comment by mrharicot** [comment]
@a-jahani @shubhamwagh the simplest way to fix it is to change `range(4)` to `range(8)` in https://github.com/mrharicot/monodepth/blob/master/monodepth_model.py#L355 and https://github.com/mrharicot/monodepth/blob/master/monodepth_model.py#L356

I will not update this code anymore as I mostly keeping it around so that people can compare with the CVPR submission.

a new pytorch code with an improved method will be online in the next month or so in a separate repo

**Comment by abhiagwl4262** [comment]
when I try to add them up in a single line of list comprehension, I get an error stating the sizes for gradients are different. This is somewhat expected since the gradient_x is of size (H, W-1) while the gradient_y is of size (H-1, W) as a consequence of your function to get the X and Y image gradients.

Any solution for this ??

**Comment by drmaj** [comment]
@mrharicot Why the switch to pytorch? I'm curious about switching as well.

**Comment by mrharicot** [comment]
It is very easy to use and you can iterate and try things extremely fast

**Comment by mrharicot** [comment]
@abhiagwl4262 because we are using tf.abs you can simply compute the mean separately for both or you can also crop each of them so that the sizes matches (H-1, W-1)

**Comment by roboticsbala** [comment]
> @a-jahani @shubhamwagh the simplest way to fix it is to change `range(4)` to `range(8)` in https://github.com/mrharicot/monodepth/blob/master/monodepth_model.py#L355 and https://github.com/mrharicot/monodepth/blob/master/monodepth_model.py#L356
> from my experience, it shouldn't really impact the results
> 
> I will not update this code anymore as I mostly keeping it around so that people can compare with the CVPR submission.
> 
> a new pytorch code with an improved method will be online in the next month or so in a separate repo

Hi @mrharicot, is the new pytorch code that you mentioned going to be an implementation of the same monodepth model or the latest model from "Digging into Self-Supervised Monocular Depth Estimation"?
